### PR TITLE
Add kernel printf util for debugging.

### PR DESCRIPTION
**Pull request by chengjunlu**
Add the printf for the Triton kernel for Intel GPU. It is useful in debugging the codegen with the printing information inside the kernel.

1. Move the llPrint method outside the PrintOpConversion to global
2. Add kernel_printf util for debugging with printing.

### CODE REVIEW

**Comment by chengjunlu** [comment]
@whitneywhtsang , @etiotto 
Please help to review.
It is useful in debugging the code with the printing information inside kernel generated.

**Review by whitneywhtsang — COMMENTED** [review]
Can `tl.device_print` be used to print information inside kernels?

**Comment by etiotto** [comment]
> Can `tl.device_print` be used to print information inside kernels?

Correct. We have the lowering for `tl.device_print` to a SPIRV function. We also have lowering for several other Triton print operators (static_print, ...).

**Comment by chengjunlu** [comment]
> Can `tl.device_print` be used to print information inside kernels?

The `tl.device_print` is the work group level ops which means the semantic is SIMD. I can double check that.

The debug printf is a simple SIMT utils which can be used in code generation at phase ttgir -> LLVM IR.

I can double check that.

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 1]
```diff
@@ -448,5 +448,84 @@ Value addStringToModule(Location loc, ConversionPatternRewriter &rewriter,
...
+    Type newType;
+    Value newArg;
+    std::tie(newType, newArg) = promoteValue(rewriter, arg);
+    operands.push_back(newArg);
+  }
+  call(funcOp, operands);
+}
+
+// Returns a Value for the format string, which you can reuse.
+Value KernelPrintf(ConversionPatternRewriter &rewriter, StringRef msg, ValueRange args) {
```
Comment (etiotto) [inline lib/Conversion/TritonGPUToLLVM/Utility.cpp:518]: This function exist in https://github.com/intel/intel-xpu-backend-for-triton/blob/64a4e507df38840e2edfeac85ad34f18299801d2/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp#L392

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 2]
```diff
@@ -448,5 +448,84 @@ Value addStringToModule(Location loc, ConversionPatternRewriter &rewriter,
...
+  SmallVector<Value> operands;
+  operands.push_back(msg);
+  for (auto arg : args) {
+    Type newType;
+    Value newArg;
+    std::tie(newType, newArg) = promoteValue(rewriter, arg);
+    operands.push_back(newArg);
+  }
+  call(funcOp, operands);
+}
```
[the remark this answers is not in the document]
Comment (chengjunlu) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:515]: I think it is good. I will change it.

**Comment by chengjunlu** [comment]
> > Can `tl.device_print` be used to print information inside kernels?
> 
> The `tl.device_print` is the work group level ops which means the semantic is SIMD. I can double check that.
> 
> The debug printf is a simple SIMT utils which can be used in code generation at phase ttgir -> LLVM IR. This utils is helpful for debugging the codegen for checking the coordinate value of each thread for the different layout encoding.
> 
> I can double check whether the `tl.device_print`.

The `tl.device_print` is mapped to the `triton::PrintOp`. It is a high-level SIMD operation.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp [hunk 3]
```diff
@@ -386,6 +386,62 @@ Value addStringToModule(Location loc, ConversionPatternRewriter &rewriter,
...
+
+  SmallVector<Value> operands;
+  operands.push_back(msg);
+  for (auto arg : args) {
+    operands.push_back(arg);
+  }
+  call(funcOp, operands);
+}
+
+// Returns a Value for the format string, which you can reuse.
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.cpp:432]: remove the comment here, as it is already added in the header file.
```suggestion
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp [hunk 4]
```diff
@@ -37,7 +37,8 @@ struct PrintOpConversion
       llvm::raw_string_ostream os(formatStr);
       os << "pid (" << getFormatSubstr(pid[0]) << ", "
          << getFormatSubstr(pid[1]) << ", " << getFormatSubstr(pid[2]) << ")%s";
-      llPrintf(formatStr, {pid[0], pid[1], pid[2], prefixStr}, rewriter);
+      mlir::LLVM::utils::llPrintf(rewriter, formatStr,
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp:40]: can we use the defined `kernel_printf` in this file?

File: third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp [hunk 5]
```diff
@@ -211,121 +213,6 @@ struct PrintOpConversion
     assert(false && "not supported type");
     return "";
   }
-
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp:214]: Note: some of these functions are moved to Utility.cpp.

File: third_party/intel/lib/TritonIntelGPUToLLVM/Utility.h [hunk 6]
```diff
@@ -142,6 +142,10 @@ using namespace mlir::triton;
 #define i32_arr_attr(...) rewriter.getI32ArrayAttr({__VA_ARGS__})
 #define i64_arr_attr(...) rewriter.getI64ArrayAttr({__VA_ARGS__})
 
+// Kernel printf
+#define kernel_printf(fmt, ...)                                                \
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/Utility.h:146]: I rather not have this macro at all and use the `llPrintf` functions defined below.

File: third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp (hunk 4 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp:40]: I think we should just remove the macro and use the llPrintf uitility functions directly.

File: third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp (hunk 5 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp:215]: Added the note.

File: third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp (hunk 4 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PrintOpToLLVM.cpp:41]: `kernel_printf` is removed.

**Review by whitneywhtsang — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Pull request merged**